File: apps/admin-console/services/knowledge.ts
import { request } from '@/services/http';

export type KnowledgeDocumentStatus = 'draft' | 'published' | 'archived';

export interface KnowledgeDocumentSummary {
  id: string;
  title: string;
  category?: string;
  tags: string[];
  version?: string;
  status: KnowledgeDocumentStatus;
  chunk_count: number;
  updated_at: string;
  created_by?: string;
  metadata?: Record<string, unknown>;
}

export interface KnowledgeDocumentListParams {
  page?: number;
  page_size?: number;
  keyword?: string;
  tags?: string[];
  status?: KnowledgeDocumentStatus | 'all';
}

export interface KnowledgeDocumentListResponse {
  documents: KnowledgeDocumentSummary[];
  total: number;
  page: number;
  page_size: number;
}

export interface KnowledgeUploadDocument {
  text: string;
  metadata: Record<string, unknown>;
}

export interface KnowledgeUploadPayload {
  documents: KnowledgeUploadDocument[];
  collection_name?: string;
  chunk_size?: number;
  chunk_overlap?: number;
}

export interface KnowledgeUploadResponse {
  success: boolean;
  inserted_count: number;
  collection_name: string;
  message: string;
}

export interface KnowledgeSearchParams {
  query: string;
  top_k?: number;
  collection_name?: string;
}

export interface KnowledgeSearchResult {
  text: string;
  score: number;
  metadata: Record<string, unknown>;
}

export interface KnowledgeSearchResponse {
  results: KnowledgeSearchResult[];
  query: string;
  total_results: number;
}

function buildQueryString(params: Record<string, unknown>) {
  const searchParams = new URLSearchParams();

  Object.entries(params).forEach(([key, value]) => {
    if (value === undefined || value === null || value === '') {
      return;
    }

    if (Array.isArray(value)) {
      value.forEach((item) => searchParams.append(key, String(item)));
      return;
    }

    searchParams.set(key, String(value));
  });

  const queryString = searchParams.toString();
  return queryString ? `?${queryString}` : '';
}

export async function listKnowledgeDocuments(
  params: KnowledgeDocumentListParams,
): Promise<KnowledgeDocumentListResponse> {
  const query = buildQueryString({
    page: params.page,
    page_size: params.page_size,
    keyword: params.keyword,
    status: params.status === 'all' ? undefined : params.status,
    tags: params.tags,
  });

  return request<KnowledgeDocumentListResponse>(`/api/v1/knowledge/documents${query}`);
}

export async function uploadKnowledgeDocuments(
  payload: KnowledgeUploadPayload,
): Promise<KnowledgeUploadResponse> {
  return request<KnowledgeUploadResponse>('/api/v1/knowledge/upsert', {
    method: 'POST',
    body: JSON.stringify({
      collection_name: payload.collection_name ?? 'knowledge_base',
      chunk_size: payload.chunk_size ?? 500,
      chunk_overlap: payload.chunk_overlap ?? 50,
      documents: payload.documents,
    }),
  });
}

export async function searchKnowledge(
  params: KnowledgeSearchParams,
): Promise<KnowledgeSearchResponse> {
  const query = buildQueryString({
    query: params.query,
    top_k: params.top_k,
    collection_name: params.collection_name,
  });

  return request<KnowledgeSearchResponse>(`/api/v1/knowledge/search${query}`);
}

File: src/models/knowledge.py
"""
知识库数据模型
"""

from datetime import datetime

from pydantic import BaseModel, Field


class DocumentChunk(BaseModel):
    """文档切片"""

    text: str = Field(..., max_length=10000, description="文档文本内容")
    metadata: dict = Field(default_factory=dict, description="文档元数据")


class KnowledgeUpsertRequest(BaseModel):
    """知识库上传请求"""

    documents: list[DocumentChunk] = Field(..., min_length=0, max_length=100)
    collection_name: str = Field(default="knowledge_base")
    chunk_size: int = Field(default=500, ge=100, le=2000)
    chunk_overlap: int = Field(default=50, ge=0, le=500)


class KnowledgeUpsertResponse(BaseModel):
    """知识库上传响应"""

    success: bool
    inserted_count: int
    collection_name: str
    message: str


class SearchResult(BaseModel):
    """搜索结果"""

    text: str
    score: float = Field(..., ge=0.0, le=1.0, description="相似度分数")
    metadata: dict


class KnowledgeSearchRequest(BaseModel):
    """知识库搜索请求"""

    query: str
    top_k: int = Field(default=3, ge=1, le=10)
    collection_name: str = Field(default="knowledge_base")


class KnowledgeSearchResponse(BaseModel):
    """知识库搜索响应"""

    results: list[SearchResult]
    query: str
    total_results: int


class KnowledgeDocumentSummary(BaseModel):
    """知识库文档概要信息"""

    id: str = Field(..., description="文档唯一标识")
    title: str = Field(..., description="文档标题")
    category: str | None = Field(default=None, description="所属分类")
    tags: list[str] = Field(default_factory=list, description="标签列表")
    version: str | None = Field(default=None, description="版本信息")
    status: str = Field(default="draft", description="文档状态")
    chunk_count: int = Field(default=1, ge=1, description="向量切片数量")
    updated_at: datetime = Field(..., description="最近更新时间")
    created_by: str | None = Field(default=None, description="创建人")
    metadata: dict = Field(default_factory=dict, description="原始元数据")


class KnowledgeDocumentListResponse(BaseModel):
    """知识库文档分页响应"""

    documents: list[KnowledgeDocumentSummary]
    total: int = Field(..., ge=0, description="文档总数")
    page: int = Field(..., ge=1, description="当前页码")
    page_size: int = Field(..., ge=1, description="每页数量")

